### ISSUE TITLE
Get rid of « Combine JavaScript files » option

### ISSUE THREAD

**Issue by GeekPress** [wp-media/wp-rocket#4023]
**Is your feature request related to a problem? Please describe.**
The proposal is to remove the feature in WP Rocket to combine JavaScript files. This will also result in removing the “Combine JavaScript files” option from WP Rocket UI as well.

**Describe the solution you'd like**
Further details related to this improvement are available [here](https://www.notion.so/wpmedia/5d9809fbfff74a30a0a6e2ab72498fa4?v=791d9573e7c34c8f87fa031d27da0b70&p=9bc993a2457545dd84bfb1a2d60868d8&pm=s).

**Comment by jazir555** [comment]
Please do not do this. My site breaks without being able to combine js. 

I don't understand why you are constantly removing features. I benefit from this feature, it doesn't improve my site to have it removed. You can simply leave the warning as is, I don't understand how this benefits people to remove a useful feature. Yes it may not be useful for everyone, but it is useful for me. How does it hurt to leave the feature in? Just don't work on it anymore. It causes more dev work by you guys to remove it.

Contrary to popular wisdom regarding http/2, my performance is actually better with combine js on. This holds for every optimization plugin i've tested.

**Comment by GeekPress** [comment]
>  My site breaks without being able to combine js.

@jazir5 You have a problem you need to fix. If you need to combine your JS files to fix something you have a serious problem with your website.

> I don't understand how this benefits people to remove a useful feature.

Everything is explained in the why section.

You can also check this post for more detailed reasons:
https://wpjohnny.com/why-you-shouldnt-combine-css-js-performance-reasons/ 

> How does it hurt to leave the feature in? 

We already explained this in the why section.

_Combine JS is one of the features that create common issues for our customers. A very common one is the cache dir size issue where combining inline JS into the combined leading to huge combined files that fills up servers. Getting rid of this feature will put an end to those issues for good._

**Comment by jazir555** [comment]
This doesn't negate the fact that I've tested it with and without combine,
and my speeds are better than without it enabled. You can say that with all
the evidence it should be faster on an http/2 site because it can download
the files in parallel, but I assure you, after testing extensively, it
performs better with combine on.

I'm not really sure why you guys are so opposed to having options,
currently the feature is still in the plugin and it's opt-in. There is no
dev harm to leaving it in, people choose whether to use it or not. While my
site may be a special case, there are legitimate use cases for combine js
if it benefits site speed.

**Comment by ghost** [comment]
We agree with @jazir5!

Stop removing **user-controllable features** and incorporating them into your plugin as Automated features or Default settings. You are taking away users' control of what they need to optimize their website. 

Another thing to keep in mind is that when WP Rocket conflicts with another plugin (e.g., SG Optimizer) -- with user-controllable features -- we can manually disable overlapping features that may cause website issues.

We don't need a lecture on what's best for our website to justify your removal of user-controllable features. Just give us (and BRING BACK) the options you have taken away. If this is not addressed and remedied soon, we will be forced to stop renewing our subscription with WP Rocket.

Do you really want to help? Here's what you can do:

(1) Bring back the structure of WPR 3.8.8 (where we can Combine JS Files AND Delay JS Execution).
(2) Convert your Automated features and Default settings to user-controllable features (i.e., toggles) as listed in [this WPR article](https://docs.wp-rocket.me/article/59-getting-started).
(3) After completing (1) and (2) above, simply add the RUCSS module you introduced with WPR 3.9.X

Oh, and have you noticed? Ever since WP Rocket started incorporating default features into its plugin, WP Rocket's Knowledgebase (KB) has grown exponentially to address and fix issues that many of your customers are experiencing. Not good!

If WP Rocket has some common sense, they will LISTEN to their customers and ACT accordingly. We have dropped many plugins that followed your "new philosophy" and found new ones that are much better (i.e., user-controllable, lightweight, and with beautiful UIs). For example, we replaced iThemes Security with SG Security, and Yoast SEO with Rank Math SEO. 

Until recently, we recommended WP Rocket to EVERYONE, now WE DON'T and WILL NOT for the above reasons.

If possible, kindly share this feedback with your team leader, Ian, and other decision makers. At this point, it appears "the tail is wagging the dog," not the other way around as it should be.

We look forward to the above changes. 

Thank you.

**Comment by androidacy-user** [comment]
#4269 before this i think. There is no parallel downloading without Push, and there is no Push without appropriate headers.

**Comment by androidacy-user** [comment]
> This doesn't negate the fact that I've tested it with and without combine, and my speeds are better than without it enabled. You can say that with all the evidence it should be faster on an http/2 site because it can download the files in parallel, but I assure you, after testing extensively, it performs better with combine on. I'm not really sure why you guys are so opposed to having options, currently the feature is still in the plugin and it's opt-in. There is no dev harm to leaving it in, people choose whether to use it or not. While my site may be a special case, there are legitimate use cases for combine js if it benefits site speed.

In  3+ years of running a (WordPress) site, and about a year of using WP Rocket, combining JS has led to nothing but hard to pinpoint problems. As said here already, if you have a situation where this is necessary, you have bigger issues.

Besides, the big performance hit from JS usually isn't just the downloading of the file - it's the parsing and execution. And a bigger file is going to take exponentially longer to download, parse, and execute, then just including the necessary small JS file(s).

Another hit with combining is that you load all JS on all pages, even if one or two files aren't needed. This leads to unused JS, which wastes time by needing to be downloaded and parsed just to never be executed.

TL;DR: combine JS is not a solution, its just a flimsy band-aid on a much bigger problem.

**Comment by jazir555** [comment]
> > This doesn't negate the fact that I've tested it with and without combine, and my speeds are better than without it enabled. You can say that with all the evidence it should be faster on an http/2 site because it can download the files in parallel, but I assure you, after testing extensively, it performs better with combine on. I'm not really sure why you guys are so opposed to having options, currently the feature is still in the plugin and it's opt-in. There is no dev harm to leaving it in, people choose whether to use it or not. While my site may be a special case, there are legitimate use cases for combine js if it benefits site speed.
> 
> In 3+ years of running a (WordPress) site, and about a year of using WP Rocket, combining JS has led to nothing but hard to pinpoint problems. As said here already, if you have a situation where this is necessary, you have bigger issues.
> 
> Besides, the big performance hit from JS usually isn't just the downloading of the file - it's the parsing and execution. And a bigger file is going to take exponentially longer to download, parse, and execute, then just including the necessary small JS file(s).
> 
> Another hit with combining is that you load all JS on all pages, even if one or two files aren't needed. This leads to unused JS, which wastes time by needing to be downloaded and parsed just to never be executed.
> 
> TL;DR: combine JS is not a solution, its just a flimsy band-aid on a much bigger problem.

That may be the case for you, but that doesn't mean the option should be removed. I have better performance with it enabled. It's the reason it's an _option_. It's _optional_. If it doesn't work better for you, don't enable it.

**Comment by androidacy-user** [comment]
> We agree with @jazir5!
Adding a billion toggles just because a handful of users want something is no solution either, and no way to run, well, anything.

You don't aim to please everyone, you aim to piss off the least amount of users. Which is whats being done here.

"Many users" do not use and like this feature but "many users" do have things inexplicably broken because someone who thought they, unlike the @wp-media team thats been doing this for much longer and in a professional capacity, knew all there is to know about website optimization, and suggested it.

**Comment by androidacy-user** [comment]
At the end of the day, you are two users vs the tens of thousands they have. Losing your business means nothing if it improves the experience for the vast majority of users.

**Comment by jazir555** [comment]
> At the end of the day, you are two users vs the tens of thousands they have. Losing your business means nothing if it improves the experience for the vast majority of users.

No one is forcing you to toggle the option on. There is already a warning cautioning people to reconsider using it. There is zero reason to remove it. It has been in the plugin for a long time, and it benefits _some_ people. That's the entire point of options. If you don't like it, **don't use it.**

**Comment by androidacy-user** [comment]
> > At the end of the day, you are two users vs the tens of thousands they have. Losing your business means nothing if it improves the experience for the vast majority of users.
> 
> No one is forcing you to toggle the option on. There is already a warning cautioning people to reconsider using it. There is zero reason to remove it. It has been in the plugin for a long time, and it benefits _some_ people. That's the entire point of options. If you don't like it, **don't use it.**

Spoken like a true user, eh.

Another mantra to live by: if few or no one uses it, or it breaks things for a lot of people, it's extra bloat to the codebase and will require unnecessary maintenance, and thus should be removed.

**Comment by androidacy-user** [comment]
It's usually not as simple as "just leave it in".

They've already said there's conflicts.

Every piece of code, every function, every option requires continuous maintenance and adds extra code to the already massive codebase.

When a future update breaks it, they have to fix it, just for the couple users that still use it, because those users threw a fit. Does that sound fair to the devs? Maintaining a feature very few use, at no gain to them or the majority of their users?

Codebases must always be updated, cared for, and maintained, and old features removed, or they will get mired in legacy code and be unable to cater to current needs. It's the thing you accept when using any software.

You are _always_ free to fork and maintain that feature yourself (thats the beauty of open source), or use a different plugin, but as it gives no benefit to the majority of users or the developers of this project  (and has the facts to back that up), it is likely to be removed anyway.

**Comment by jazir555** [comment]
> > > At the end of the day, you are two users vs the tens of thousands they have. Losing your business means nothing if it improves the experience for the vast majority of users.
> > 
> > 
> > No one is forcing you to toggle the option on. There is already a warning cautioning people to reconsider using it. There is zero reason to remove it. It has been in the plugin for a long time, and it benefits _some_ people. That's the entire point of options. If you don't like it, **don't use it.**
> 
> Spoken like a true user, eh.
> 
> Another mantra to live by: if few or no one uses it, or it breaks things for a lot of people, it's extra bloat to the codebase and will require unnecessary maintenance, and thus should be removed.

This plugin _is for the users_. That's the entire point of the product, literally. If it helps people, and people have already bought the plugin, removing features _they paid for_ is wrong. This plugin is meant to benefit the users, not the devs.

**Comment by androidacy-user** [comment]
That _two people_ use.

Get something straight right now: two users do NOT dictate the direction of ANY company. Removing a feature that **most** find useless or breaking is not going to hurt them, their business, or development. Keeping will hurt the latter.

Once again, you fail to see beyond what you want as of this exact moment. 

It's not about just you.

WP Rocket has tens of thousands of users, and yet, for all your talk of "this effects some people", i've only seen two saying the removal of this feature would impact them.

Another side note: you did **not** pay for any specific feature. You paid for the product "as-is", meaning the devs have a right to remove, add, change, or otherwise modify any part of the software as they see fit, or discontinue the plugin entirely, at their sole discretion.

Stuff your disrespect and entitlement. It has no place on a GitHub issue tracker.

**Comment by jazir555** [comment]
>WP Rocket has tens of thousands of users, and yet, for all your talk of "this effects some people", i've only seen two saying the removal of this feature would impact them.

Not everybody in the world has a github, github is going to attract a small minority of wp-rocket users. If you were to post this on facebook, for instance, where it would have significantly more visibility, you would most certainly get a different reaction. How about they make a user poll before just making the decision? How do _you_ know how many people use the feature? Your intuition?

>Stuff your disrespect and entitlement.

I could say the same. You don't seem to care at all for people that spent money with a feature that is advertised and included in the documentation. You make plenty of assumptions about the features usage, provide some evidence of its lack of use. Do you have any data to back that up?

**Comment by jbma** [comment]
Hi all, we've locked and limited the conversation to collaborators.
First, thank's a lot for providing feedback and contributing to our product developpement.

Our Github repo is the developers work tool, having a lot of not related (to developpement) content make it harder for us to work and collaborate in a good way. That's the reason we've locked it.

On a more global way, our product developpement vision is aligned with [WordPress philosophy ](https://wordpress.org/about/philosophy/) and especially : 
- Decision not options
- Design for the majority 

This specific improvement is a pure alignement of this philosophy, to make WP Rocket more efficient and simpler.
